cxx/distributions/zero_mean_normal_test.cc

```diff
@@ -0,0 +1,105 @@
+// Apache License, Version 2.0, refer to LICENSE.txt
+
+#define BOOST_TEST_MODULE test Normal
+
+#include "distributions/zero_mean_normal.hh"
+
+#include <boost/test/included/unit_test.hpp>
+
+#include "util_math.hh"
+namespace bm = boost::math;
+namespace tt = boost::test_tools;
+
+BOOST_AUTO_TEST_CASE(simple) {
+  std::mt19937 prng;
+  ZeroMeanNormal nd(&prng);
+
+  nd.incorporate(5.0);
+  nd.incorporate(-2.0);
+  BOOST_TEST(nd.N == 2);
+
+  nd.unincorporate(5.0);
+  nd.incorporate(7.0);
+  BOOST_TEST(nd.N == 2);
+
+  nd.unincorporate(-2.0);
+  BOOST_TEST(nd.N == 1);
+}
+
+BOOST_AUTO_TEST_CASE(no_nan_after_incorporate_unincorporate) {
+  std::mt19937 prng;
+  ZeroMeanNormal nd(&prng);
+
+  nd.incorporate(10.0);
+  nd.unincorporate(10.0);
+
+  BOOST_TEST(nd.N == 0);
+  BOOST_TEST(!std::isnan(nd.var));
+}
+
+BOOST_AUTO_TEST_CASE(logp_before_incorporate) {
+  std::mt19937 prng;
+  ZeroMeanNormal nd(&prng);
+
+  BOOST_TEST(nd.logp(6.0) == -5.4563792395895785, tt::tolerance(1e-6));
```

Comment:
I'd replace these tests with marginalization tests (see normal_test.cc to test against a boost computed logp and logp_score via approximate integration).

I find the testing exact value tests brittle and have the possibility of hiding bugs, and strongly prefer marginalization tests.

Reply:
I agree that marginalization tests are great.  They are also a lot more work.  My preference would be to have these minimal tests now, and then add the marginalization tests in a later pull request.

Reply:
The recommendation here is to copy the tests from normal_test.cc. I've added marginalization tests for logp and logp_score. Those should work the exact same here.

Reply:
https://github.com/probcomp/hierarchical-irm/blob/a03fa0bf1c77fdb473878dbffb9a13a8eb1c01d5/cxx/distributions/normal_test.cc#L36 and the following test_posterior_pred should work here (with some renaming of distributions).

Reply:
So I tried your recommendation and it didn't work.  The latest commit of this pull request has the test_log_prob and it fails.

This shouldn't be surprising, because as previously noted, this distribution has different marginal likelihood and predictive posterior functions than the Normal.  For example, when we do add a marginalization test for this distribution, it will only need to do a single integral over variances rather than the double integral done in normal_test.cc.